Only allow 'unique' and 'well-known' symbols

acutmore opened the pull request:
Made a start on what the spec text changes could be if _registered_ and _well-known_ symbols were not allowed in `WeakMap`, `WeakSet`, `WeakRef` and `FinalizationRegistry`

Closes #21

acutmore commented on spec.emu:
```diff
@@ -38,7 +89,7 @@ contributors:
       1. Let _entries_ be the List that is _M_.[[WeakMapData]].
       1. <del>If Type(_key_) is not Object, return *false*.</del>
-      1. <ins>If IsObjectOrSymbol(_key_) is *false*, return *false*.</ins>
+      1. <ins>If ! %CanBeHeldWeakly%(_key_) is *false*, return *false*.</ins>
```
Do we think this is OK to direct use the intrinsic, or should I split it up into a core AO, and then the separate intrinsic which is just a wrapper around the AO?

ljharb replied:
if you're directly using it, you'd have to `Call()` it, since it's a JS function object. it's probably better to have a common AO that both reference.

acutmore replied:
Thanks @ljharb, I should have gone with my gut! I've updated so there is a common AO to use.

acutmore commented on spec.emu:
```diff
@@ -13,22 +13,73 @@ contributors:
+          </tr>
+          <tr>
+            <td>
+              <ins>%CanBeHeldWeakly%</ins>
```
We create this as a new intrinsic so it can be shared across `WeakMap.isValidKey`, `WeakSet.isValidValue`, `WeakRef.isValidTarget` and `FinalizationRegistry.isValidTarget`.

acutmore commented on spec.emu:
```diff
@@ -13,22 +13,73 @@ contributors:
+    1. If Type(_argument_) is Symbol, then
+      1. For each element _e_ of the GlobalSymbolRegistry List (see <emu-xref href="#sec-symbol.for"></emu-xref>), do
+        1. If SameValue(_e_.[[Symbol]], _argument_) is *true*, return *false*.
+      1. Let _wellKnownSymbols_ be the symbols listed in <emu-xref href="#sec-well-known-symbols"></emu-xref>.
```
I'm referencing the section rather than the table, because I got a `ecmarkup` error saying it couldn't find the `xref` when referencing the table.

ljharb commented on spec.emu:
```diff
@@ -99,7 +150,13 @@ contributors:
+  <h1>Modifications to WeakSet</h1>
+
+  <emu-clause id="sec-weakset.isValidValue">
+    <h1>WeakMap.isValidValue( _value_ )</h1>
```
Technically this predicate is answering if it's a valid *key* - everything's a valid value.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -13,22 +13,73 @@ contributors:
+  </emu-clause>
+</emu-clause>
+<emu-clause id="sec-%canbeheldweakly%">
+  <h1>%CanBeHeldWeakly%( _argument_ )</h1>
```
We need to explicitly define `.name`, otherwise it will probably be missing or `"%CanBeHeldWeakly%"` (I never read the part of the spec that automatically defines function names). Or you could define it as an "exposed function" directly, similarly to how `Array.prototype.values` is defined normally and then `Array.prototype.@@iterator` "copies" it.

nicolo-ribaudo replied:
I should have read 10 lines below before commenting :upside_down_face:

bathos commented on spec.emu:
```diff
@@ -13,22 +13,73 @@ contributors:
+      1. For each element _e_ of the GlobalSymbolRegistry List (see <emu-xref href="#sec-symbol.for"></emu-xref>), do
+        1. If SameValue(_e_.[[Symbol]], _argument_) is *true*, return *false*.
+      1. Let _wellKnownSymbols_ be the symbols listed in <emu-xref href="#sec-well-known-symbols"></emu-xref>.
+      1. If _wellKnownSymbols_ contains _argument_, return *false*.
```
This might be relevant here: https://github.com/tc39/test262/issues/3420.

The %Intl%.[[Fallback]] symbol is specified as a unique symbol associated with the realm. The text here would consider it to be unforgeable, which is theoretically true and will be in some implementations, e.g. Spidermonkey.

V8 however implemented the fallback symbol with the single-identity-across-realms characteristic of a well-known symbol. Though technically incorrect, it seems this may have been a deliberate decision and the discussion about which behavior is desired did not appear conclusive. This proposal might force the issue of getting engines / the Intl spec to agree on the true nature of this abomination.

ljharb replied:
Thats something that needs to either be changed in the spec, or, v8 needs to change their implementation, and it should be discussed in plenary regardless of this PR/proposal.

acutmore replied:
If I have understood that issue, this is not an issue for this proposal in practice. Because the symbol value itself is not exposed anywhere, so it wouldn't be possible to try and use it in a `WeakSet` for example. So the spec would say that it is not well-known (cross realm) so _can_ be used in a `WeakSet`, V8 implement it cross-realm so it is well-known so if they didn't allow it in a `WeakSet` that would be incorrect, but that incorrect behaviour can't be triggered in any way.

Please do correct anything I have misunderstood.

bathos replied:
Yes. Though as you mentioned in another thread, the status of well-knowns as “forgeable” is pretty iffy\*. If it’s decided that well-known symbols aren’t considered forgeable, then this outlier-in-practice thing is moot anyway.

\* It’s strictly impossible to “prove” either conclusion in all possible scenarios from an ES code runtime POV, so while the spec describes them in terms that consistently support the unforgeable conclusion, I understand that this isn’t cut and dry since if a tree falls in the forest...

acutmore commented:
> It is my understanding that for forward web compatibility we would prefer to actually name the exposed predicate something like `isUnforgeable`.

Thanks @rricard - good point that this was still being discussed. I've created an issue https://github.com/tc39/proposal-symbols-as-weakmap-keys/issues/24

acutmore commented:
Based on the most recently conversations in the Record & Tuple monthly call - I have updated the PR:

- removed the previously included static predicates (e.g. `WeakMap.isValidKey`)
- widened the `CanBeHeldWeakly` AO to also return true for _well-known symbols_ by only returning false for symbols that are _registered_.

ljharb commented on spec.emu:
```diff
@@ -71,7 +115,7 @@ contributors:
       1. Let _entries_ be the List that is _M_.[[WeakMapData]].
       1. <del>If Type(_key_) is not Object, return *false*.</del>
-      1. <ins>If IsObjectOrSymbol(_key_) is *false*, return *false*.</ins>
+      1. <ins>If ! CanBeHeldWeakly(_key_) is *false*, return *false*.</ins>
```
```suggestion
      1. <ins>If CanBeHeldWeakly(_key_) is *false*, return *false*.</ins>
```
same throughout, since this AO doesn’t return a completion record

acutmore replied:
thanks!

bathos commented:
> also return true for well-known symbols by only returning false for symbols that are registered.

@acutmore I’d thought about this (way too much) and came to the same conclusion, but wasn’t confident about my ability to articulate the case coherently. It’s really cool to see that this is where you’ve landed!
